clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProvider.java

```diff
@@ -120,36 +122,48 @@ private boolean refreshCredentials() throws IOException {
     final ZeebeClientCredentials fetchedCredentials = fetchCredentials();
     credentialsCache.put(endpoint, fetchedCredentials).writeCache();
 
-    if (fetchedCredentials.equals(credentials)) {
-      return false;
+    if (credentials == null || !credentials.isValid() || !fetchedCredentials.equals(credentials)) {
+      credentials = fetchedCredentials;
+      LOG.debug("Refreshed credentials.");
+
+      return true;
     }
 
-    credentials = fetchedCredentials;
-    LOG.debug("Refreshed credentials.");
-    return true;
+    return false;
   }
 
-  private static String createJsonPayload(final OAuthCredentialsProviderBuilder builder)
-      throws JsonProcessingException {
+  private static String createParams(final OAuthCredentialsProviderBuilder builder) {
     final Map<String, String> payload = new HashMap<>();
     payload.put("client_id", builder.getClientId());
     payload.put("client_secret", builder.getClientSecret());
     payload.put("audience", builder.getAudience());
     payload.put("grant_type", "client_credentials");
 
-    return JSON_MAPPER.writeValueAsString(payload);
+    final StringBuilder sb = new StringBuilder();
+    payload.forEach((key, value) -> sb.append(String.format("%s=%s&", encode(key), encode(value))));
+    sb.deleteCharAt(sb.length() - 1);
+    return sb.toString();
+  }
+
+  private static String encode(final String param) {
+    try {
+      return URLEncoder.encode(param, StandardCharsets.UTF_8.name());
+    } catch (UnsupportedEncodingException e) {
+      throw new UncheckedIOException("Failed while encoding OAuth request parameters: ", e);
+    }
   }
 
   private ZeebeClientCredentials fetchCredentials() throws IOException {
     final HttpURLConnection connection =
-        (HttpURLConnection) authorizationServerUrl.openConnection();
+        (HttpURLConnection)
+            new URL(String.format("%s?%s", authorizationServerUrl, queryParams)).openConnection();
```

Comment:
Do we need to add the queryParams as query params if we pass them as body as well?

Reply:
No, that's a mistake. Thanks